sdks/java/io/google-cloud-platform/src/main/java/org/apache/beam/sdk/io/gcp/pubsub/PubsubJsonClient.java

```diff
@@ -136,20 +136,23 @@ public int publish(TopicPath topic, List<OutgoingMessage> outgoingMessages)
     for (OutgoingMessage outgoingMessage : outgoingMessages) {
       PubsubMessage pubsubMessage = new PubsubMessage().encodeData(outgoingMessage.elementBytes);
 
-      Map<String, String> attributes = outgoingMessage.attributes;
-      if ((timestampAttribute != null || idAttribute != null) && attributes == null) {
-        attributes = new TreeMap<>();
+      Map<String, String> attributes = null;
+      if ((timestampAttribute != null || idAttribute != null)) {
+        if (outgoingMessage.attributes == null) {
+          attributes = new TreeMap<>();
+        } else {
+          attributes = new TreeMap<>(outgoingMessage.attributes);
```

Comment:
Why are you doing this inside the outer if? I think this means that you will drop all extra attributes if timestampAttribute and idAttribute are not set.

Reply:
If timestampAttribute and idAttribute are not set, the new attributes map will be null, and the call to pubsubMessage.setAttributes below won't be reached, retaining the original attributes map. I kept the original code structure intact, opting to just fix the bug in this change.

Reply:
not sure I understand - what if the users sets other attributes?

Reply:
If the user sets other attributes and the timestamp/id attributes are null, then those original attributes will not be overwritten because pubsubMessage.setAttributes is never called. That call mutates the message, overwriting the attributes map that already exists and could contain user-specified attributes with the new attributes map constructed here.

Reply:
Working on a test for this case so you don't have to take my word for it. :)

Reply:
You were correct. I misunderstood how the messages were being created and missed a second bug in the original code.

Since I had to fix the bug anyway, I also took the opportunity to refactor the publish method.